File: sdk/rust/src/staking_info.rs
use hedera_proto::services;
use serde::Serialize;
use serde_with::{
    serde_as,
    TimestampNanoSeconds,
};
use time::OffsetDateTime;

use crate::protobuf::ToProtobuf;
use crate::{
    AccountId,
    FromProtobuf,
    Hbar,
};

// todo(sr): is this right?
/// Info related to account/contract staking settings.
#[serde_as]
#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct StakingInfo {
    /// If `true`, the contract declines receiving a staking reward. The default value is `false`.
    pub decline_staking_reward: bool,

    /// The staking period during which either the staking settings for this account or contract changed (such as starting
    /// staking or changing staked_node_id) or the most recent reward was earned, whichever is later. If this account or contract
    /// is not currently staked to a node, then this field is not set.
    #[serde_as(as = "Option<TimestampNanoSeconds>")]
    pub stake_period_start: Option<OffsetDateTime>,

    /// The amount in `Hbar` that will be received in the next reward situation.
    pub pending_reward: Hbar,

    /// The total of balance of all accounts staked to this account or contract.
    pub staked_to_me: Hbar,

    /// The account to which this account or contract is staking.
    pub staked_account_id: Option<AccountId>,

    /// The ID of the node this account or contract is staked to.
    pub staked_node_id: Option<u64>,
}

impl FromProtobuf<services::StakingInfo> for StakingInfo {
    fn from_protobuf(pb: services::StakingInfo) -> crate::Result<Self> {
        let (staked_account_id, staked_node_id) = match pb.staked_id {
            Some(services::staking_info::StakedId::StakedAccountId(id)) => {
                (Some(AccountId::from_protobuf(id)?), None)
            }

            Some(services::staking_info::StakedId::StakedNodeId(id)) => (None, Some(id as u64)),

            None => (None, None),
        };

        Ok(Self {
            decline_staking_reward: pb.decline_reward,
            stake_period_start: pb.stake_period_start.map(Into::into),
            pending_reward: Hbar::from_tinybars(pb.pending_reward),
            staked_to_me: Hbar::from_tinybars(pb.staked_to_me),
            staked_account_id,
            staked_node_id,
        })
    }
}

impl ToProtobuf for StakingInfo {
    type Protobuf = services::StakingInfo;

    fn to_protobuf(&self) -> Self::Protobuf {
        services::StakingInfo {
            decline_reward: self.decline_staking_reward,
            stake_period_start: self.stake_period_start.map(Into::into),
            pending_reward: self.pending_reward.to_tinybars(),
            staked_to_me: self.staked_to_me.to_tinybars(),
            staked_id: self
                .staked_node_id
                .map(|it| it as i64)
                .map(services::staking_info::StakedId::StakedNodeId)
                .or_else(|| {
                    self.staked_account_id
                        .as_ref()
                        .map(ToProtobuf::to_protobuf)
                        .map(services::staking_info::StakedId::StakedAccountId)
                }),
        }
    }
}

File: sdk/swift/Sources/Hedera/StakingInfo.swift
/*
 * ‌
 * Hedera Swift SDK
 * ​
 * Copyright (C) 2022 - 2023 Hedera Hashgraph, LLC
 * ​
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ‍
 */

import CHedera
import Foundation
import HederaProtobufs

public struct StakingInfo: Codable {
    internal init(
        declineStakingReward: Bool,
        stakePeriodStart: Timestamp?,
        pendingReward: Hbar,
        stakedToMe: Hbar,
        stakedAccountId: AccountId?,
        stakedNodeId: UInt64?
    ) {
        self.declineStakingReward = declineStakingReward
        self.stakePeriodStart = stakePeriodStart
        self.pendingReward = pendingReward
        self.stakedToMe = stakedToMe
        self.stakedAccountId = stakedAccountId
        self.stakedNodeId = stakedNodeId
    }

    /// If true, the contract declines receiving a staking reward. The default value is false.
    public let declineStakingReward: Bool

    /// The staking period during which either the staking settings for this account or contract changed
    /// (such as starting staking or changing staked_node_id)
    /// or the most recent reward was earned, whichever is later.
    /// If this account or contract is not currently staked to a node, then this field is not set.
    public let stakePeriodStart: Timestamp?

    /// The amount in Hbar that will be received in the next reward situation.
    public let pendingReward: Hbar

    /// The total of balance of all accounts staked to this account or contract.
    public let stakedToMe: Hbar

    /// The account to which this account or contract is staking.
    public let stakedAccountId: AccountId?

    /// The ID of the node this account or contract is staked to.
    public let stakedNodeId: UInt64?

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)

        declineStakingReward = try container.decode(Bool.self, forKey: .declineStakingReward)
        stakePeriodStart = try container.decodeIfPresent(Timestamp.self, forKey: .stakePeriodStart)
        pendingReward = try container.decode(Hbar.self, forKey: .pendingReward)
        stakedToMe = try container.decode(Hbar.self, forKey: .stakedToMe)
        stakedAccountId = try container.decodeIfPresent(AccountId.self, forKey: .stakedAccountId)
        stakedNodeId = try container.decodeIfPresent(UInt64.self, forKey: .stakedNodeId)
    }

    public static func fromBytes(_ bytes: Data) throws -> Self {
        try Self(fromProtobufBytes: bytes)
    }

    public func toBytes() -> Data {
        toProtobufBytes()
    }
}

extension StakingInfo: TryProtobufCodable {
    internal typealias Protobuf = Proto_StakingInfo

    internal init(fromProtobuf proto: Protobuf) throws {
        let stakePeriodStart = proto.hasStakePeriodStart ? proto.stakePeriodStart : nil
        let stakedAccountId: Proto_AccountID?
        let stakedNodeId: UInt64?
        switch proto.stakedID {
        case .none:
            stakedAccountId = nil
            stakedNodeId = nil
        case .some(.stakedNodeID(let nodeId)):
            stakedAccountId = nil
            stakedNodeId = UInt64(nodeId)
        case .some(.stakedAccountID(let accountId)):
            stakedAccountId = accountId
            stakedNodeId = nil
        }

        self.init(
            declineStakingReward: proto.declineReward,
            stakePeriodStart: .fromProtobuf(stakePeriodStart),
            pendingReward: Hbar.fromTinybars(proto.pendingReward),
            stakedToMe: Hbar.fromTinybars(proto.stakedToMe),
            stakedAccountId: try .fromProtobuf(stakedAccountId),
            stakedNodeId: stakedNodeId
        )
    }

    func toProtobuf() -> Protobuf {
        .with { proto in
            proto.declineReward = declineStakingReward
            stakePeriodStart?.toProtobufInto(&proto.stakePeriodStart)
            proto.pendingReward = pendingReward.toTinybars()
            proto.stakedToMe = stakedToMe.toTinybars()

            stakedAccountId?.toProtobufInto(&proto.stakedAccountID)

            // node ID wins, so it goes last.
            if let nodeId = stakedNodeId {
                proto.stakedNodeID = Int64(nodeId)
            }
        }
    }
}
